core/src/test/java/com/graphhopper/routing/weighting/DirectedRoutingTest.java

```diff
@@ -0,0 +1,323 @@
+package com.graphhopper.routing.weighting;
+
+import com.graphhopper.Repeat;
+import com.graphhopper.RepeatRule;
+import com.graphhopper.routing.*;
+import com.graphhopper.routing.ch.PrepareContractionHierarchies;
+import com.graphhopper.routing.lm.PrepareLandmarks;
+import com.graphhopper.routing.util.*;
+import com.graphhopper.storage.*;
+import com.graphhopper.storage.index.LocationIndex;
+import com.graphhopper.storage.index.LocationIndexTree;
+import com.graphhopper.storage.index.QueryResult;
+import com.graphhopper.util.CHEdgeIteratorState;
+import com.graphhopper.util.EdgeExplorer;
+import com.graphhopper.util.EdgeIterator;
+import com.graphhopper.util.GHUtility;
+import com.graphhopper.util.shapes.BBox;
+import com.graphhopper.util.shapes.GHPoint;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.util.*;
+
+import static com.graphhopper.util.EdgeIterator.ANY_EDGE;
+import static com.graphhopper.util.EdgeIterator.NO_EDGE;
+import static com.graphhopper.util.Parameters.Algorithms.ASTAR_BI;
+import static com.graphhopper.util.Parameters.Algorithms.DIJKSTRA_BI;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+/**
+ * This test makes sure the different bidirectional routing algorithms correctly implement restrictions of the source/
+ * target edges, by comparing with {@link DijkstraBidirectionRef}
+ *
+ * @author easbar
+ * @see DirectedBidirectionalDijkstraTest
+ */
+@RunWith(Parameterized.class)
+public class DirectedRoutingTest {
+    private final Algo algo;
+    private final double uTurnCosts;
+    private final boolean prepareCH;
+    private final boolean prepareLM;
+    private Directory dir;
+    private GraphHopperStorage graph;
+    private CHGraph chGraph;
+    private CarFlagEncoder encoder;
+    private TurnCostExtension turnCostExtension;
+    private int maxTurnCosts;
+    private Weighting weighting;
+    private EncodingManager encodingManager;
+    private PrepareContractionHierarchies pch;
+    private PrepareLandmarks lm;
+
+    @Rule
+    public RepeatRule repeatRule = new RepeatRule();
+
+    @Parameterized.Parameters(name = "{0}, u-turn-costs: {1}")
+    public static Collection<Object[]> params() {
+        return Arrays.asList(new Object[][]{
+                {Algo.ASTAR, Double.POSITIVE_INFINITY, false, false},
+                {Algo.CH_ASTAR, Double.POSITIVE_INFINITY, true, false},
+                {Algo.CH_DIJKSTRA, Double.POSITIVE_INFINITY, true, false},
+                // todo: yields warnings and fails
+//                {Algo.LM, Double.POSITIVE_INFINITY, false, true}
```

Comment:
With landmarks the test fails sometimes, maybe its the same problem as in #1557 ? Or maybe I am using bad settings (there were also some warnings about some weights being maxed out)

Reply:
This is ugly. Wasn't #1557 only for shortest weighting and shouldn't happen here?

Reply:
Hmm, I might need your help here. I guess the test for LM fails because I am getting these warnings:

```shell
com.graphhopper.routing.lm.LandmarkStorage - landmark 1 (49.40121656807996,9.705012881122817): too many delta weights were maxed out (17/82). Factor is too small 0.003551062425503072. To fix this increase maximum in config.yml: prepare.lm.weighting: fastest|maximum=363.056352875689
```

I tried setting the maximum weight with `LandmarkStorage#setMaximumWeight` and also setting it to `-1` (automatic determination of maximum weight), but the warnings remain. I do not really understand what the problem is here. The random graph is created within a rather small bounding box (just the size of a village): http://tools.geofabrik.de/calc/#type=geofabrik_standard&bbox=9.7,49.41,9.71,49.4&tab=1&proj=EPSG:4326&places=2
and I checked that the edge distances are all something like 100-1000m. The speeds used are all between 10 and 120 so there should not be any unusual values for the weights (I checked and they are all somewhere between 60 and 420 (one to seven minutes).
However the `fromWeight` here: https://github.com/graphhopper/graphhopper/blob/022697da8dfc9ab6e6537fbc7b81efc55ac2e142/core/src/main/java/com/graphhopper/routing/lm/LandmarkStorage.java#L554 seems to be really large (something like 30-80k, so several hours up to a full day) and because of this the `delta` in the same method is often below `DELTA_MIN`, which eventually leads to the warnings I mentioned above.

What I do not understand is how `fromWeight` can get so large ? I would expect in such a small area no matter where the landmarks are the weights should be much smaller ?

Reply:
Thanks, will investigate! This sounds like a bug in LM code.

Reply:
Hmmh, there are a few confusing things in the code like 16 bits for the precision although we should have 18 for the "from" and 13 for the "delta", plus old comments making it hard to "re-understand" this. It will take me a bit more time before I can understand this problem here.

Reply:
Ok, thanks for taking a look! Maybe we can leave this test commented out for this PR ? The problem seems to be unrelated to the source/target edge restrictions that are new here ?

Reply:
Will let you know when I know if it is indeed unrelated. (But maybe it is related as the preparation and also parts of the algo are node based)

Reply:
I tried looking into this a bit and  found failing cases also without the new feature discussed here (and also for node-based routing). See https://github.com/graphhopper/graphhopper/issues/1687

Reply:
Thanks a lot!!